File: pr-1012/sphinx/snippets/python/using/cudaq/platform/sample_async_remote.py
# ============================================================================ #
# Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
# [Begin Documentation]
import cudaq

# Number of remote QPUs to launch
num_qpus = 2
cudaq.set_target("remote-mqpu",
                 remote_execution=True,
                 backend="tensornet",
                 auto_launch=str(num_qpus))
target = cudaq.get_target()
print("Number of QPUs:", target.num_qpus())

kernel, runtime_param = cudaq.make_kernel(int)
qubits = kernel.qalloc(runtime_param)
# Place qubits in superposition state.
kernel.h(qubits)
# Measure.
kernel.mz(qubits)

count_futures = []
for qpu in range(num_qpus):
    count_futures.append(cudaq.sample_async(kernel, 4 + qpu, qpu_id=qpu))

for counts in count_futures:
    print(counts.get())

File: pr-1169/sphinx/snippets/cpp/using/cudaq/platform/sample_async_remote.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// Compile and run with:
// ```
// nvq++ sample_async.cpp -o sample_async.x --target remote-mqpu
// --remote-mqpu-auto-launch 2
// && ./sample_async.x
// ```
#include <cudaq.h>

int main() {
  // [Begin Documentation]
  auto [kernelToBeSampled, runtimeParam] = cudaq::make_kernel<int>();
  auto q = kernelToBeSampled.qalloc(runtimeParam);
  kernelToBeSampled.h(q);
  kernelToBeSampled.mz(q);

  // Get the quantum_platform singleton
  auto &platform = cudaq::get_platform();

  // Query the number of QPUs in the system
  // The number of QPUs is equal to the number of auto-launch server instances.
  auto num_qpus = platform.num_qpus();
  printf("Number of QPUs: %zu\n", num_qpus);
  // We will launch asynchronous sampling tasks
  // and will store the results immediately as a future
  // we can query at some later point.
  // Each QPU (indexed by an unique Id) is associated with a remote REST server.
  std::vector<cudaq::async_sample_result> countFutures;
  for (std::size_t i = 0; i < num_qpus; i++) {
    countFutures.emplace_back(cudaq::sample_async(
        /*qpuId=*/i, kernelToBeSampled, /*runtimeParam=*/5));
  }

  //
  // Go do other work, asynchronous execution of sample tasks on-going
  //

  // Get the results, note future::get() will kick off a wait
  // if the results are not yet available.
  for (auto &counts : countFutures) {
    counts.get().dump();
  }
  // [End Documentation]
  return 0;
}
